### ISSUE TITLE
CLI para o signer

### ISSUE THREAD

**Issue by denydias** [demoiselle/signer#304]
Olá! Primeiramente, quero parabenizar todos os colaboradores deste projeto. Uma implementação de referência para o PBAD do porte do Demoiselle Signer é importância vital para sociedade como um todo.

Tenho integrado o Signer em diversos projetos. Quando o projeto é Java nativo, o Maven resolve o tema. Mas há projetos em muitas outras linguagens, e aí a integração começa a se complicar, chegando aos casos de uso extremo onde torna-se impossível usar o Demoiselle Signer.

Por isso eu pergunto: existe no roadmap o desenvolvimento de um CLI (Command Line Interface) para o Signer? Algo como:

```
java -jar /path/to/signer/DemoiselleSigner.jar
```

Isso facilitaria muito a integração do Signer em praticamente qualquer projeto! Obviamente, o tratamento dais saídas de `stdout` e `stderr` continuam sob responsabilidade do integrador.

Este é um exemplo do que me refiro: [JSignPdf Quick Start Guide](http://jsignpdf.sourceforge.net/uploads/JSignPdf.pdf), seção `9 Command line (batch mode)`, pág. 26.

De antemão agradeço qualquer feedback dos colaboradores do projeto.

**Comment by alexandre-mbm** [comment]
@denydias, é possível usá-lo com NEOiD em Android?

**Comment by denydias** [comment]
> @denydias, é possível usá-lo com NEOiD em Android?

Não. O  JSignPdf utiliza certificados armazenados de acordo com PKCS#12.

**Comment by alexandre-mbm** [comment]
@denydias 

> Tenho integrado o Signer em diversos projetos. Quando o projeto é Java nativo, o Maven e código resolvem o tema. Mas há projetos em muitas outras linguagens onde a integração começa a se complicar, chegando aos casos de uso extremo onde torna-se impossível usar o Demoiselle Signer.

Referi-me ao Demoiselle Signer.

**Comment by denydias** [comment]
> Referi-me ao Demoiselle Signer.

Aí é melhor você consultar a [documentação](https://www.frameworkdemoiselle.gov.br/v3/signer/docs/) do projeto.

**Comment by esaito** [comment]
O Assinador SERPRO, pode fazer algo nesse sentido, através de chamadas websocket. https://www.serpro.gov.br/assinador-digital/

**Comment by denydias** [comment]
> O Assinador SERPRO, pode fazer algo nesse sentido, através de chamadas websocket. https://www.serpro.gov.br/assinador-digital/

Pode, mas muito aquém das possibilidades de uma linha de comando. Por exemplo:

- O Assinador SERPRO é distribuído somente para Windows, macOS e distribuições com kernel Linux Debian/Ubuntu, Fedora e openSUSE. Especialmente nos ambientes em nuvem, existem uma infinidade de cenários onde esses pacotes simplesmente não podem ser utilizados;
- Exige a instalação e execução em tempo integral do Assinador SERPRO em ao menos uma máquina, que pode ser a mesma onde a aplicação está em execução ou outra qualquer, o que implica em configuração adicional para a possibilidade de uso do websocket;
- Exige uso de portas TCP (local ou remota);
- Exige uso de um navegador;
- Não pode carregar um certificado sob demanda (fora do keystore do próprio Assinador);
- Documentação pobre.

Particularmente (leia-se: minha opinião), não gosto do modelo que o Serpro desenvolve e distribui seus aplicativos. Sendo uma empresa estatal, fornecer o Assinador SERPRO sem uma licença de código aberto e livre robusta como ocorre com o Demoiselle Signer invalida completamente a sua possibilidade de uso integrado ou embutido em outras aplicações.

Uma CLI para o Demoiselle Signer me parece uma solução muito mais segura, transparente, robusta e perene para possibilitar assinaturas sob demanda em qualquer ambiente.
